File: node/src/services/data/datasource/inmemory.datasource.ts
import { Ok } from '../../../utils/result.util'
import { IDataSource } from './datasource'
import _ from 'lodash'

export class InMemoryDataSource implements IDataSource {
  data: Map<string, unknown>
  constructor() {
    this.data = new Map()
  }
  has(key: string): Promise<Result<boolean, Error>> {
    return Promise.resolve(Ok(this.data.has(key)))
  }

  async init(): Promise<IDataSource> {
    return this
  }

  async getAll<T>(keyPrefix: string): Promise<Result<T[], Error>> {
    keyPrefix = keyPrefix.endsWith('*') ? keyPrefix.substring(0, keyPrefix.length - 1) : keyPrefix
    const result = [...this.data.entries()]
      .filter(([key]) => key.startsWith(keyPrefix))
      .map(([, value]) => _.cloneDeep<T>(value as T))
    return Ok(result)
  }

  async set<T>(key: string, data: T): Promise<Result<T, Error>> {
    const clone = _.cloneDeep(data)
    this.data.set(key, clone)
    return Ok(data)
  }

  async get<T>(key: string): Promise<Result<T, Error>> {
    const keyExists = this.data.has(key)
    if (!keyExists) return new Error(`Key ${key} not found`)
    const value = this.data.get(key) as T
    const clone = _.cloneDeep(value)
    return Ok(clone)
  }

  async delete(key: string): Promise<Result<boolean, Error>> {
    const keyExists = this.data.has(key)
    if (!keyExists) return new Error(`Key ${key} not found`)
    this.data.delete(key)
    return Ok(true)
  }
}

File: go/pkg/data/inmemory_datasource.go
package data

import "errors"

type inmemoryDataSource struct {
	store map[string]string
}

func NewInMemoryDataSource() DataSource {
	return &inmemoryDataSource{
		store: make(map[string]string),
	}
}

func (r *inmemoryDataSource) Init() {}

func (r *inmemoryDataSource) Has(key string) (bool, error) {
	_, ok := r.store[key]
	return ok, nil
}

func (r *inmemoryDataSource) Get(key string) (string, error) {
	value, ok := r.store[key]
	if !ok {
		return "", errors.New("key not found")
	}
	return value, nil
}

func (r *inmemoryDataSource) GetAll(prefix string) ([]string, error) {
	values := []string{}
	for key, value := range r.store {
		if key[:len(prefix)] == prefix {
			values = append(values, value)
		}
	}
	return values, nil
}

func (r *inmemoryDataSource) Set(key string, value string) (string, error) {
	r.store[key] = value
	return value, nil
}

func (r *inmemoryDataSource) Delete(key string) (string, error) {
	value, ok := r.store[key]
	if !ok {
		return "", errors.New("key not found")
	}
	delete(r.store, key)
	return value, nil
}
